design/bmc-address.md

```diff
@@ -0,0 +1,126 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# bmc-address
+
+## Status
+
+implementable
+
+## Table of Contents
+
+<!--ts-->
+   * [bmc-address](#bmc-address)
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Alternatives](#alternatives)
+      * [References](#references)
+
+<!-- Added by: dhellmann, at: 2019-03-04T09:33-0500 -->
+
+<!--te-->
+
+## Summary
+
+This document explains the way users will provide the network location
+of the management controller on a host.
+
+## Motivation
+
+We need to document how we are going to specify the location of a BMC,
+including how to tell its type, IP address, port number, and
+potentially the path.
+
+### Goals
+
+1. To agree on an address specification system, including assumptions
+   that can be made when parts of an address are left out.
+
+### Non-Goals
+
+1. To list the scheme to use for every type of controller.
+2. To specify the user interface for entering address information.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+For each BMC, we need to know the type, IP, port, and optionally the
+path to use to communicate with the controller.
+
+We can collect all of this information using a single "address" field,
+if we allow that field to contain partial or full URLs.
+
+For each type of controller, we can assume we know the URL scheme used
+to communicate with it (HTTP, HTTPS, etc.). Therefore the scheme in a
+URL provided by the user is redundant, and we can use that portion of
+the URL to specify the controller type. For example:
+
+    ipmi://192.168.111.1
```

Comment:
This looks nice and consistent. I'm a bit worried about splitting the redfish URL. The unfortunate bit of redfish is that it takes two parameters. E.g. `https://192.168.111.1/redfish/v1/Systems/42` is split into:
```
redfish_address = https://192.168.111.1/redfish/v1
redfish_system_id = /redfish/v1/Systems/42
```
I'm not sure what to do about it. Probably just look for `/redfish/v1` bit, but it's not guaranteed to be there.. I'd like to summon @etingof to comment on it.

Reply:
This is definitely not something we want to require users to do. Is that prefix consistent? Even if there are a few different ones, we could encode all of them. Is "redfish" a "type" in the sense that IPMI is a type here? If so, I think we'd want users to enter something like redfish://IP/redfish/v1/Systems/42 and then we would turn that into the values Ironic needs.

Reply:
I think it's not unreasonable to expect this, but I'm not sure that the Redfish standard actually dictates the `/redfish/v1/` bit. We can probably ignore it for now (like we ignore IPMI bridging and other craziness).

Reply:
Redfish standard calls `/redfish` part as the *root service*, the `v1` part - *protocol version*. The standard does not dictate BMC to expose the root service and the protocol version in exactly `/redfish/v1` form.

I think `redfish://IP/redfish/v1/Systems/42` form should suffice, though may be I'd also include the original schema because https is not strictly required. For example, can it be `redfish+https://host/redfish/v1/Systems/42'?

Reply:
I added the redfish example, thanks for pointing that out as a special case.